### ISSUE TITLE
p2p: Handshake PoW

### ISSUE THREAD

**Issue by ebuchman** [tendermint/tendermint#1045]
Handshake should check that the peers ID satisfies some PoW.

Handshake should also include a proof of work challenge.

**Comment by ebuchman** [comment]
Design: 

- Should be implemented like SecretConnection - a wrapper for a net.Conn that performs the handshake
- The listening peer issues a random 32 byte challenge; the dialing peer responds with a 32 byte nonce.
- The big endian encoding of the 32-byte HASH( challenge || nonce) is less than 2^(256-difficulty) - 1.
- Provide config options for `min_in_difficulty` (minimum difficulty a peer must beat for us to accept the connection) and `max_out_difficulty` (determines how much proof of work we'll do).

This gives us maximum flexibility, but the min_in isn't published explicitly. New nodes will just have to start increasing their max_out_difficulty until they manage to connect to peers. 

Alternatively, we could include "min_in" in the challenge,  but we'd still need max_out to prevent malicious peers from making everyone do tons of work to connect to bad peers.

**Comment by ValarDragon** [comment]
I think `min_in` should be in the challenge. 

Is anyone working on this? If not, I can write a PR for this.

**Comment by mappum** [comment]
The general wisdom in trying out PoW tokens in the p2p world is that attackers often have more resources for generating tokens than legitimate users do, so it's ineffective at filtering out bad peers. Hashcash was invented to combat email spam but was eventually abandoned because of this.

I think it might be more effective to just locally keep a list of banned IP addresses for bad peers.

**Comment by ebuchman** [comment]
So, it can't really filter out bad peers, and that isn't really the intention here - rather it's to increase the cost of opening many connections. That said, I guess it's not really clear that that's useful either, since we can already prevent multiple connections from the same IP and we need to keep a list of banned IP addresses anyways (#1614). Maybe we shouldn't move forward with it unless there's really good justification, as it will also add complexity to the configuration for users who don't need it at all

**Comment by ValarDragon** [comment]
I feel like a PoW connection may still help if we face an adversary with a botnet trying to DoS nodes all across the network. (I.e. adversary with many IP's, trying to attack everyone), as a PoW in effect limits the rate at which they can spam. (Though the PoW I had begun to implement should be adapted such that only the one trying to initiate the connection must do the work)

I agree though that we can definitely table this, and investigate later if the need arises.

**Comment by jlandrews** [comment]
I agree that we will want to introduce more robust networking level spam / DDOS prevention, but I feel like a more complete system of paying for bandwidth might be more useful long term. Using POW as a sort of proof of burned resources to access my resources doesn't actually give me any incentive to accept your request, just gives me some assurance that you can't spam with impunity. A system which actually involves a transfer of value from requester to provider would give better spam prevention IMO, and would help address things like cosmos/cosmos-sdk#1541 in an incentive compatible way. That sort of solution is obviously a much larger endeavor though, and not an immediate priority.

**Comment by ValarDragon** [comment]
I think the difference is that doing something like that (similar to Orchid) is perhaps a multi-year project, whereas proofs of work are an easy to implement solution. 

The point of a proof of work is not to incentivize accepting a request, but instead to purely act as DDOS protection. (Before getting you to do a <relatively> expensive secret connection handshake). And I disagree, verifying a proof of work will always be _significantly_ easier than verifying a single digital signature. (Most digital signatures hash the input anyway, ed25519 has two hashes), so even paying for connections wouldn't fix this problem for spam mitigation, since the signature could just be wrong. (Unless you propose causing a cost per byte sent to be incurred when communicating over the internet, which is well beyond the scope of a software project)

**Comment by jlandrews** [comment]
I guess if the POW is meant to prevent the cost of the accepting the connections then you're right that general network incentives don't help with that. I was seeing POW as more of a way of putting a cost on the connection to prevent wasteful use of connections, and in that framework, I was suggesting we make a longer term plan to address this fully in the future.

**Comment by ValarDragon** [comment]
Not really sure what you mean by wasteful use of connections. The intent here is purely to avoid being DDOS'd by people trying to connect to you, which happens by limiting the amount of work an attacker can make you do. This achieved by a PoW by only allowing one connection per IP, and then limiting the rate at which an attacker can create connections.

**Comment by melekes** [comment]
Just for the reference:

```
S/Kademlia based IPFS identity generation:
      difficulty = <integer parameter>
      n = Node{}
      do {
        n.PubKey, n.PrivKey = PKI.genKeyPair()
        n.NodeId = hash(n.PubKey)
        p = count_preceding_zero_bits(hash(n.NodeId))
      } while (p < difficulty)
```
